src/qs8-vadd/hvx.c.in

```diff
@@ -0,0 +1,143 @@
+// Copyright 2020 Google LLC
+//
+// This source code is licensed under the BSD-style license found in the
+// LICENSE file in the root directory of this source tree.
+
+$assert DATATYPE in ["QS8", "QU8"]
+$assert BATCH_TILE % 8 == 0
+$assert BATCH_TILE >= 8
+$ABC = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+#include <assert.h>
+
+#include <hvx_hexagon_protos.h>
+#include <hexagon_protos.h>
+#include <hexagon_types.h>
+
+#include <xnnpack/vbinary.h>
+#include <xnnpack/intrinsics-polyfill.h>
+
+$XINT8_T = {"QS8": "int8_t", "QU8": "uint8_t"}[DATATYPE]
+
+void xnn_${DATATYPE.lower()}_vadd_minmax_ukernel__hvx_u${BATCH_TILE}(
+    size_t batch,
+    const ${XINT8_T}* input_a,
+    const ${XINT8_T}* input_b,
+    ${XINT8_T}* output,
+    const union xnn_${DATATYPE.lower()}_add_minmax_params params[restrict XNN_MIN_ELEMENTS(1)]) XNN_OOB_READS
+{
+  assert(batch != 0);
+  assert(batch % sizeof(${XINT8_T}) == 0);
+  assert(input_a != NULL);
+  assert(input_b != NULL);
+  assert(output != NULL);
+
+  const HVX_Vector vbias = *((const HVX_Vector *)(params->hvx.bias));
+  const HVX_Vector va_multiplier = *((const HVX_Vector *)(params->hvx.a_multiplier));
+  const HVX_Vector vb_multiplier = *((const HVX_Vector *)(params->hvx.b_multiplier));
+  const int32_t vshift = params->hvx.shift;
+  const HVX_Vector voutput_zero_point = *((const HVX_Vector *)(params->hvx.output_zero_point));
+  const HVX_Vector voutput_min = *((const HVX_Vector *)(params->hvx.output_min));
+  const HVX_Vector voutput_max = *((const HVX_Vector *)(params->hvx.output_max));
+  int8_t* ptr_o = output;
+
+  for (; batch >= ${BATCH_TILE} * sizeof(${XINT8_T}); batch -= ${BATCH_TILE} * sizeof(${XINT8_T})) {
+    HVX_Vector va${ABC[0]} = *((HVX_UVector*)input_a);
+    HVX_Vector vb${ABC[0]} = *((HVX_UVector*)input_b);
+    $for N in range(32, BATCH_TILE, 32):
+      HVX_Vector va${ABC[int(N/32)]} = *((HVX_UVector*)(input_a + ${N}));
+      HVX_Vector vb${ABC[int(N/32)]} = *((HVX_UVector*)(input_b + ${N}));
+    input_a += ${BATCH_TILE};
+    input_b += ${BATCH_TILE};
+
+    // unpack: 8 bit to 16 bit
+    $for N in range(0, BATCH_TILE, 32):
+      HVX_Vector va${ABC[int(N/32)]}_i16_lo = Q6_V_lo_W(Q6_Wh_vunpack_Vb(va${ABC[int(N/32)]}));
+      HVX_Vector vb${ABC[int(N/32)]}_i16_lo = Q6_V_lo_W(Q6_Wh_vunpack_Vb(vb${ABC[int(N/32)]}));
+
+    // vacc = vbias + va * va_multiplier
+    $for N in range(0, BATCH_TILE, 32):
+      HVX_Vector va${ABC[int(N/32)]}_mul_e = Q6_Vw_vmpyio_VwVh(va_multiplier, Q6_Vh_vshuffe_VhVh(va${ABC[int(N/32)]}_i16_lo, va${ABC[int(N/32)]}_i16_lo));
```

Comment:
although batch_tile will help, consider putting the $for N in range(0, BATCH_TILE, 32):
around a single loop that does va and vb, since the instructions will have no dependencies

    // vacc = vbias + va * va_multiplier + vb * vb_multiplier
    $for N in range(0, BATCH_TILE, 32):
      HVX_Vector va${ABC[int(N/32)]}_mul_e = Q6_Vw_vmpyio_VwVh(va_multiplier, Q6_Vh_vshuffe_VhVh(va${ABC[int(N/32)]}_i16_lo, va${ABC[int(N/32)]}_i16_lo));
      HVX_Vector vb${ABC[int(N/32)]}_mul_e = Q6_Vw_vmpyio_VwVh(vb_multiplier, Q6_Vh_vshuffe_VhVh(vb${ABC[int(N/32)]}_i16_lo, vb${ABC[int(N/32)]}_i16_lo));
      HVX_Vector va${ABC[int(N/32)]}_mul_o = Q6_Vw_vmpyio_VwVh(va_multiplier, va${ABC[int(N/32)]}_i16_lo);
      HVX_Vector vb${ABC[int(N/32)]}_mul_o = Q6_Vw_vmpyio_VwVh(vb_multiplier, vb${ABC[int(N/32)]}_i16_lo);
      HVX_VectorPair va${ABC[int(N/32)]}_mul = Q6_W_vshuff_VVR(va${ABC[int(N/32)]}_mul_o, va${ABC[int(N/32)]}_mul_e, -4);
      HVX_VectorPair vb${ABC[int(N/32)]}_mul = Q6_W_vshuff_VVR(vb${ABC[int(N/32)]}_mul_o, vb${ABC[int(N/32)]}_mul_e, -4);
      HVX_Vector vacc${ABC[int(N/32)]} = Q6_Vw_vadd_VwVw(vbias, Q6_V_lo_W(va${ABC[int(N/32)]}_mul));
      vacc${ABC[int(N/32)]} = Q6_Vw_vadd_VwVw(vacc${ABC[int(N/32)]}, Q6_V_lo_W(vb${ABC[int(N/32)]}_mul));

Reply:
Done